lib/response-handler.js

```diff
@@ -18,53 +18,120 @@ module.exports = (bluebird, error, HttpLinkHeader, logger, ShelfError, URI) => {
+     * This only requires that a status property exists.
+     *
+     * @see {@link https://visionmedia.github.io/superagent/#error-handling}
+     *
+     * @param {superagent~Error} reason
+     * @return {shelfLib~ShelfError}
+     */
+    function createErrorForHttp(reason) {
+        var errorCode, errorCodeMap, mappedCode, statusCode;
+
+        errorCodeMap = {
+            404: error.NOT_FOUND,
+            500: error.INTERNAL_SERVER_ERROR
+        };
+        statusCode = reason.status;
+        mappedCode = Object.keys(errorCodeMap).find((code) => {
+            code = parseInt(code, 10);
+
+            return code === statusCode;
         });
-        logger.debug(`Got error code: ${errorCode}`);
-
-        errorMessage = error.UNKNOWN;
-
-        /* This could be in multiple formats at this point.
-         * We only want to deal with the errors if a body
-         * exists because that is how we have identified that
-         * the error has come from the response instead of an
-         * internal error like a socket timeout.
-         */
-        if (errorResponse.body) {
-            errorMessage = errorResponse.body.message || error.UNKNOWN;
+        errorCode = errorCodeMap[mappedCode] || error.UNKNOWN;
+
+        return new ShelfError(`Got status code of "${statusCode}" with body "${reason.response.text}"`, errorCode);
+    }
+
+    /**
+     * Creates a ShelfError from a TCP error. These are
+     * things like connection reset and socket timeout.
+     * This requires the "code" property to exist.
+     *
+     * @see {@link https://visionmedia.github.io/superagent/#error-handling}
+     *
+     * @param {superagent~Error} reason
+     * @return {shelfLib~ShelfError}
+     */
+    function createErrorForTcp(reason) {
+        var err;
+
+        if (reason.code === error.TIMEOUT) {
+            err = new ShelfError("A socket timeout occured", error.TIMEOUT);
+        } else if (reason.code === error.COULD_NOT_RESOLVE_HOST) {
+            err = new ShelfError("Could not resolve the host provided", error.COULD_NOT_RESOLVE_HOST);
+        } else if (reason.code === error.CONNECTION_REFUSED) {
+            err = new ShelfError("Unable to make a connection to the server. This is likely because the port we are trying to connect to is not listening", error.CONNECTION_REFUSED);
+        } else {
+            err = new ShelfError(reason.message, error.UNKNOWN);
         }
 
-        if (errorCode) {
-            return new ShelfError(errorMessage, error[errorCode]);
+        return err;
+    }
+
+
+    /**
+     * Maps error codes from the response to the internal error object.
+     * Returns the corresponding error.
+     *
+     * @see {@link https://visionmedia.github.io/superagent/#error-handling}
+     *
+     * @param {superagent~Error} reason
+     * @return {ShelfError}
+     */
+    function createErrorForResponse(reason) {
+        var body, err;
+
+        logger.debug("In responseHandler.handleErrorResponse()");
```

Comment:
Haha I am lazy. I was like well that is a response so I will put it in the response handler